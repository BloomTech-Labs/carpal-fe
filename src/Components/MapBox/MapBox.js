import React, { useState, useEffect } from "react";
import ReactMapGL, { Marker } from "react-map-gl";
import { connect } from "react-redux";
import Logo from "./../../img/Logo.png";

import "./MapBox.scss";
import "mapbox-gl/dist/mapbox-gl.css";
import { config } from "dotenv";
config();

const mapboxAPI = process.env.REACT_APP_MAPBOX_TOKEN;

function MapBox(props) {
    const [viewport, setViewport] = useState({
        latitude: props.favoriteLocation[0].latitude,
        longitude: props.favoriteLocation[0].longitude,
        zoom: 15,
        width: "100%",
        height: "200px",
        position: "center"
    });
    //State for keeping track of the Markers long/lat
    const [marker, setMarker] = useState({
        latitude: props.favoriteLocation[0].latitude,
        longitude: props.favoriteLocation[0].longitude
    });
    useEffect(() => {
        navigator.geolocation.getCurrentPosition(getUserLocation);
    }, [viewport]);

    const getUserLocation = (position) => {
        var crd = position.coords;
        setViewport({
            ...viewport,
            latitude: crd.latitude,
            longitude: crd.longitude
        });
        //When we Get users location we set the marker to the users current location
        setMarker({
            latitude: crd.latitude,
            longitude: crd.longitude
        });
    };

    const handleDragEnd = (event) => {
        setViewport({
            ...viewport,
            longitude: event.lngLat[0],
            latitude: event.lngLat[1]
        });
        //Update Users location on dragend
        setMarker({
            longitude: event.lngLat[0],
            latitude: event.lngLat[1]
        });
    };
    return (
        <div role="map-wrapper" className="map">
            <ReactMapGL
                {...viewport}
                mapboxApiAccessToken={mapboxAPI}
                onViewportChange={(viewport) => {
                    setViewport(viewport);
                }}
            >
                <Marker
                    latitude={marker.latitude}
                    longitude={marker.longitude}
                    offsetLeft={-10}
                    offsetTop={-10}
                    draggable={true}
                    onDragEnd={handleDragEnd}
                >
                    <img
                        src={`${Logo}`}
                        alt="marker"
                        style={{
                            width: "20px",
                            height: "20px",
                            borderRadius: "50%"
                        }}
                    />
                </Marker>
            </ReactMapGL>
        </div>
    );
}

const mapStateToProps = (state) => {
    return {
        favoriteLocation: state.user.user.favoriteLocation
    };
};

export default connect(mapStateToProps, {})(MapBox);
