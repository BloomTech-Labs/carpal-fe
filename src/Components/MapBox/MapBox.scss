.map {
    width: 100%;

    .mapboxgl-map {
        border-radius: 20px;
    }
}
