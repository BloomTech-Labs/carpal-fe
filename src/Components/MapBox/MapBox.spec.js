import React from "react";
import * as rtl from "@testing-library/react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import MapBox from "./MapBox";
import "@testing-library/jest-dom/extend-expect";

afterEach(rtl.cleanup);

// mock mapbox map, not in use currently
// jest.mock("mapbox-gl/dist/mapbox-gl", () => ({
//     Map: () => ({})
// }));

//mock navigator's getCurrentPosition function
const mockGeolocation = {
    getCurrentPosition: jest.fn().mockImplementationOnce((vals) => vals)
};

//assign geolocation the mockGeolocation obj
navigator.geolocation = mockGeolocation;

//mock redux store
let store;
const mockstore = configureStore([thunk]);

beforeEach(() => {
    //user obj in the mockstore.
    store = mockstore({
        user: {
            user: {
                first_name: "Daniel",
                last_name: "Martin",
                phone_number: 55555555,
                email: "dang@carpal.com",
                is_driver: true,
                hobbies: [],
                audio_love: [],
                audio_hate: [],
                favoriteLocation: [
                    {
                        latitude: 32.715736,
                        longitude: -117.161087
                    }
                ]
            }
        }
    });
});
// test for component displays to the client.
describe("display mapbox", () => {
    test("Displays component", () => {
        const wrapper = rtl.render(
            <Provider store={store}>
                <MapBox />
            </Provider>
        );

        const label = wrapper.getByRole(/map-wrapper/i);
        expect(label).toBeInTheDocument();
        expect(label).toBeVisible();
    });
    //snapshot testing props being passed to the component. OH SNAP
    test("matches snapshot", () => {
        const tree = renderer.create(
            <Provider store={store}>
                <MapBox />
            </Provider>
        );
        expect(tree).toMatchSnapshot();
        // returning obj
        console.log(tree.toJSON());
    });
    test("Navigator function working", () => {
        expect(navigator.geolocation.getCurrentPosition).toHaveBeenCalled();
    });
});
