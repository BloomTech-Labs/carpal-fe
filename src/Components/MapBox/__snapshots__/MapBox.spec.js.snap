// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display mapbox matches snapshot 1`] = `
<div
  className="map"
  role="map-wrapper"
>
  <div
    style={
      Object {
        "cursor": "grab",
        "height": 0,
        "position": "relative",
        "width": 0,
      }
    }
  >
    <div
      style={
        Object {
          "height": "100%",
          "position": "relative",
          "width": "100%",
        }
      }
    >
      <div
        className=""
        style={
          Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "absolute",
            "visibility": "inherit",
            "width": "100%",
          }
        }
      />
      <div
        style={
          Object {
            "height": 0,
            "overflow": "visible",
            "width": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;
